File: UniversityApi.API/Repositories/StudentsRepository.cs
using Microsoft.EntityFrameworkCore;
using UniversityApi.API.Contracts;
using UniversityApi.API.DataBase;
using UniversityApi.API.DataBase.Entities;

namespace UniversityApi.API.Repositories
{
    public class StudentsRepository : GenericRepository<EntityStudent>, IStudentsRepository
    {
        public StudentsRepository(UniversityApiDbContext dbContext) : base(dbContext)
        {
        }

        public override async Task<EntityStudent> GetAsync(Guid id)
        {
            return await _context.Set<EntityStudent>()
                .Include(s => s.Courses)
                    .ThenInclude(c => c.Professor)
                .SingleOrDefaultAsync(s => s.EntityPersonID == id);
        }

        public override async Task<List<EntityStudent>> GetAllAsync()
        {
            return await _context.Set<EntityStudent>()
                .Include(s => s.Courses)
                    .ThenInclude(c => c.Professor)
                .ToListAsync();
        }

        public async Task<Guid> UpdateWithCoursesAsync(EntityStudent updatedStudent, IEnumerable<Guid> coursesIds)
        {
            var updatedCourses = coursesIds
                .Select(id => _context.Courses.Find(id))
                .ToList();
            var studentToUpdate = _context.Set<EntityStudent>()
                .Include(s => s.Courses)
                .SingleOrDefault(s => s.EntityPersonID == updatedStudent.EntityPersonID);
            studentToUpdate.Courses = updatedCourses;
            _context.Update(studentToUpdate);
            await _context.SaveChangesAsync();
            return studentToUpdate.EntityPersonID;
        }

        public async Task DeleteStudentsCourseAsync(Guid studentId, Guid courseId)
        {
            var student = await GetAsync(studentId);
            var course = await _context.Courses.FindAsync(courseId);
            student.Courses.Remove(course);
            await _context.SaveChangesAsync();
        }

        public async Task<bool> IndexIsOccupied(string index)
        {
            return await _context.Set<EntityStudent>().AnyAsync(s => s.Index.Equals(index));
        }

        public async Task<EntityStudent> AddWithCoursesAsync(EntityStudent createdStudent, IEnumerable<Guid> coursesIds)
        {
            var studentsCourses = coursesIds
                .Select(id => _context.Courses.Find(id))
                .ToList();
            createdStudent.Courses = studentsCourses;
            var result = await _context.AddAsync(createdStudent);
            await _context.SaveChangesAsync();
            return createdStudent;
        }
    }
}

File: FsUniversityApi/Repositories/StudentsRepository.fs
namespace FsUniversityApi.Repositories

open FsUniversityApi.Database.FsDbContext
open FsUniversityApi.Database.Entities.StudentCourseProfessor
open FsUniversityApi.Contracts
open Microsoft.EntityFrameworkCore;
open System
open System.Threading.Tasks
open System.Collections.Generic
open System.Linq

type StudentsRepository(context: FsDbContext) =
    inherit GenericRepository<Student>(context)

    member this.GetAllAsync () =
            context.Set<Student>().Include(fun s -> s.Courses).ToListAsync()

    member this.GetAsync (id : Guid) =
            context.Set<Student>().Include(fun s -> s.Courses).SingleOrDefaultAsync(fun s -> s.PersonInfoId.Equals(id))

    interface IStudentsRepository with

        member this.IndexIsOccupied (index : string) =
            context.Set<Student>().AnyAsync(fun s -> s.Index.Equals(index));

        member this.UpdateWithCoursesAsync (student : Student, coursesIds: IEnumerable<Guid>) : Task<Guid> =
            async {
                let! courses = coursesIds.Select(fun id -> context.Courses.FindAsync(id).AsTask() |> Async.AwaitTask) |> Async.Parallel
                let updated = { student with Courses = ResizeArray(courses) }
                context.Remove(student) |> ignore
                do! context.AddAsync(updated).AsTask() |> Async.AwaitTask |> Async.Ignore
                do! context.SaveChangesAsync() |> Async.AwaitTask |> Async.Ignore
                return student.PersonInfoId
            }
            |> Async.StartAsTask
            
        member this.AddWithCoursesAsync (student : Student, coursesIds: IEnumerable<Guid>) =
            async {
                let! courses = coursesIds.Select(fun id -> context.Courses.FindAsync(id).AsTask() |> Async.AwaitTask) |> Async.Parallel
                let created = { student with Courses = ResizeArray(courses) }
                do! context.AddAsync(created).AsTask() |> Async.AwaitTask |> Async.Ignore
                do! context.SaveChangesAsync() |> Async.AwaitTask |> Async.Ignore
                return student
            }
            |> Async.StartAsTask

        member this.DeleteStudentsCourseAsync (studentId: Guid, courseId: Guid) =
            async {
                let! student = context.Set<Student>().FindAsync(studentId).AsTask() |> Async.AwaitTask
                let! courseToDelete = context.Courses.FindAsync(courseId).AsTask() |> Async.AwaitTask
                let updatedCourses = student.Courses |> List.ofSeq |> List.filter (fun course -> course.CourseId <> courseToDelete.CourseId)
                let updated = { student with Courses = ResizeArray(updatedCourses) }
                context.Remove(student) |> ignore
                do! context.AddAsync(updated).AsTask() |> Async.AwaitTask |> Async.Ignore
                return! context.SaveChangesAsync() |> Async.AwaitTask
            }
            |> Async.StartAsTask :> Task
        